Added support for 7-Seg counter and improved Neopixel LED strip and effects

gustavoalara opened the pull request:
<h1 class="western">
Summary of New Features 
</h1>
<p>

</p>
<h2 class="western">Firmware Improvements 
</h2>
<p>The firmware has been refactored to support a much deeper level of
customization, especially regarding the NeoPixel LED strip.<br/>
Full
support has been added for 7-Segment Counters, and NeoPixel LED strip
functionality has been expanded: 
</p>
<p><br/>
</p>
<h3 class="western">7-Segment Counter: 
</h3>
<p>Full functionality has been implemented for whmxe-595-2 counters
based on the 74HC595 shift register, allowing life or ammo to be
displayed. 
</p>
<p>
<img width="309" height="167" alt="image" src="https://github.com/user-attachments/assets/287fdeed-858c-4de5-a4eb-434e5bd7b8c3" />

  <br clear="left"/>
</img>
</p>
<p>Additionally, the two-character message shown on the 7-segment
counter when the device is powered on is customizable. This message
is configured from the GUI. 
</p>
<p><br/>
</p>
<h3 class="western">NeoPixel LED Strip Effects: 
</h3>
<p>A complete system has been created, including life and ammo bars
with gradient colors and a dynamic effects engine (Fire, Ice, Plasma,
Beam, Knight Rider) with a 10-color palette, all configurable via
serial command. 
</p>
<p><br/>
</p>
<h3 class="western">Multi-Sector LED Strip: 
</h3>
<p>The most significant improvement is the ability to divide a single
NeoPixel LED strip into up to four functional and independent
sectors: 
</p>
<p>
</p>
<ul>
	<li><p><strong>Fixed LED Sector:</strong> A number of LEDs at the
	beginning of the strip whose color is set from the GUI and is not
	affected by any other function. Ideal for indicating the player’s
	color. 
	</p></li>
</ul>
<ul>
	<li><p><strong>Life Bar Sector:</strong> A group of LEDs functioning
	as a life bar, with a customizable color gradient. 
	</p></li>
</ul>
<ul>
	<li><p><strong>Ammo Bar Sector:</strong> A group of LEDs functioning
	as an ammo bar, also with a customizable color gradient. 
	</p></li>
</ul>
<ul>
	<li><p><strong>Effects Sector:</strong> A group of LEDs dedicated
	exclusively to displaying dynamic visual effects (Fire, Ice, etc.). 
	</p></li>
</ul>
<p>This feature is configured through new settings in the GUI,
allowing you to define the start LED and the number of LEDs for each
sector. 
</p>
<p><br/>
</p>
<h3 class="western">LED Effects Assignment to Buttons: 
</h3>
<p>It is now also possible to assign any of the above-mentioned
effects to some of the lightgun’s buttons (Trigger, Gun A, Gun B,
Gun C, Pump Action, Pedal, Alt Pedal, Start, and Select) both
On-Screen and Off-Screen, allowing a greater degree of customization
for your OpenFIRE lightgun. 
</p>
<p><br/>
</p>
<h2 class="western">GUI Improvements (OpenFIRE-App) 
</h2>
<p>To support the new firmware features, the desktop application has
been updated as follows: 
</p>
<p><br/>
</p>
<h3 class="western">Board Layout tab: 
</h3>
<p>Communication pins for the counter have been added for assignment
in the board diagram. 
</p>
<p><br/>
<img width="224" height="124" alt="image" src="https://github.com/user-attachments/assets/4b6ad35f-5838-4899-8621-1d08a37692c7" />

  <br clear="left"/>
</img>

<br/>
<br/>

</p>
<h3 class="western">NeoPixel Section in Gun Settings tab:</h3>
<p>A new group has been added within the NeoPixel section for “LED
Segment Configuration,” where you can define the start LED and the
length of the life bar, ammo bar, and effects sector.<br/>
Real-time
validation has been implemented to prevent saving configurations
where segments overlap or exceed the total number of LEDs, including
an error message showing the overlap issue. 
</p>
<p>
<img width="935" height="192" alt="image" src="https://github.com/user-attachments/assets/863f3e5d-c4d0-44c7-a56d-069846f0082c" />

<br/>
<br/>

</p>
<p>A new group has been added for assigning effects to buttons. 
</p>
<p>
<img width="915" height="343" alt="image" src="https://github.com/user-attachments/assets/38e1cfaa-6f04-4ffb-ad08-fd51beb14b18" />

<br/>
<br/>

</p>
<h3 class="western">7-Seg Counter Section in settings Tab:</h3>
<p>Activation of the 7-segment counter and the initial counter type
value has been added (modifiable via serial commands).<br/>
A new
text field has been added to enter the custom startup message for the
7-segment counter. 
</p>
<p>
<img width="948" height="105" alt="image" src="https://github.com/user-attachments/assets/741c1120-c9e5-4fba-a7b2-5f56cb2ac5fe" />

<br/>
<br/>

</p>
<h2 class="western">Serial Port Communication (Mamehooker) 
</h2>
<p>
</p>
<h3 class="western">New Mode Commands (M): 
</h3>

Command | Function
-- | --
MCx1: | Sets the 7-segment counter to Life Mode.
MCx2: | Sets the 7-segment counter to Ammo Mode.
MCx0: | Turns off the 7-segment counter.


<p>         
</p>
<h3 class="western">New Effect Commands (FX): 
</h3>
<p>All effects now follow a unified format:
<code class="western">&lt;FX{Effect}x{Color}x{State}&gt;</code> 
</p>
<p><i>Effect Codes: </i>
</p>
<p>F = Fire<br/>
I = Ice<br/>
P = Plasma<br/>
B = Beam (Flash)<br/>
K
= Knight Rider 
</p>
<p><i>Color Palette: </i>
</p>
<p>R = Red<br/>
G = Green<br/>
B = Blue<br/>
O = Orange<br/>
P =
Purple<br/>
Y = Yellow<br/>
C = Cyan<br/>
M = Magenta<br/>
W =
White<br/>
L = Lime 
</p>

<p><i>State:</i></p>
<p>1 = Activate<br/>
0 = Deactivate 
</p>

<p><strong><i><span style="font-weight: normal">Examples:</span></i></strong><br/>
Activate
Fire effect in Orange: <code class="western">&lt;FXFxOx1&gt;</code><br/>
Activate
Ice effect in Cyan: <code class="western">&lt;FXIxCx1&gt;</code><br/>
Activate
Knight Rider in Blue: <code class="western">&lt;FXKxBx1&gt;</code><br/>
Deactivate
any effect: <code class="western">&lt;FX...x0&gt;</code> (e.g.,
<code class="western">&lt;FXFxOx0&gt;</code>) 
</p>

<p><strong><i><span style="font-weight: normal">Existing Life and
Ammo Commands:</span></i></strong></p>
<p><code class="western">&lt;FDLx{nn}&gt;</code> - Value for
life<br/>
<code class="western">&lt;FDAx{nn}&gt;</code> - Value for
ammo</p>
<p><strong><i><span style="font-weight: normal">Examples:</span></i></strong></p>
<p><code class="western">&lt;FDLx45&gt;</code> - Sets life to 45 (on
the LED bar it is converted into a percentage, with the first value
received being considered 100%).</p>
<p><code class="western">&lt;FDAx99&gt;</code> - Sets ammo to 99 (on
the LED bar it is converted into a percentage, with the first value
received being considered 100%).</p>

</body>
</html>

gustavoalara committed:
Added external counter via Arduino+UART 

Added external two digits counter support controlled by an Arduino via UART communication. The code in the Arduino is the PapaGustavoKratos' code (mine) that you can find in pistolasdeluzopedia.miraheze.org

gustavoalara committed:
Added Life/Amno to counter via serial

gustavoalara committed:
Create OpenFireCounter.h

gustavoalara committed:
Create OpenFireCounter.cpp

gustavoalara committed:
Update OpenFIREDefines.h

gustavoalara committed:
Update OpenFIREcommon.h

gustavoalara committed:
Update OpenFIREmain.ino

gustavoalara committed:
Update OpenFIREserial.cpp

gustavoalara committed:
Update OpenFIREmain.ino

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.h

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

Fix inverted numbers

gustavoalara committed:
Fix inverted numbers again

gustavoalara committed:
again inverted numbers

gustavoalara committed:
and again fixing fonts on counter

gustavoalara committed:
Fixing 6 and 9

gustavoalara committed:
Let's check if all the numbers are fixed now

gustavoalara committed:
Fixing the last inverted number

gustavoalara committed:
again...

gustavoalara committed:
and again with the 4

gustavoalara committed:
Added LSB First for fonts

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Update OpenFireCounter.h

gustavoalara committed:
Update OpenFireCounter (hope last one)

gustavoalara committed:
Update OpenFireCounter.h

gustavoalara committed:
Update OpenFireCounter.cpp

gustavoalara committed:
Totally fixed

gustavoalara committed:
I'm totally idiot

gustavoalara committed:
Updated OpenFIREmain.ino Counter data from preferences

gustavoalara committed:
Updated OpenFIREmain.ino Mamehooker part

gustavoalara committed:
Update .gitmodules (changed to my updated Boards)

gustavoalara committed:
Update arduino-cli-builds.yml

gustavoalara committed:
feat: Update shared assets submodule

gustavoalara committed:
Update and rename OpenFireCounter.cpp to OpenFIREcounter.cpp

Comments to english

gustavoalara committed:
Update and rename OpenFireCounter.h to OpenFIREcounter.h

Translated comments to english

gustavoalara committed:
Change OpenFireCounter to OpenFIREcounter

Change OpenFireCounter to OpenFIREcounter for naming convention

gustavoalara committed:
Change OpenFireCounter to OpenFIREcounter

Change OpenFireCounter to OpenFIREcounter for naming convention

gustavoalara committed:
Update arduino-cli-builds.yml

gustavoalara committed:
Added support for Neopixel Life/Ammo bar, Neopixel effects (#2)

* Added Neopixel Life/Amno Bar

* Added Neopixel LED Effects

* added the new effects commands

* Improved and consolidated serial commands for effects

gustavoalara committed:
Added segments for Neopixel strand (#3)

* LED Effects and LED bar by sectors

* Remove ML command and modify FDL/FDA

Remove ML command as is not need with the new multi sector led approach and modify FDL/FDA to handle the target sector

* Modified updateNeoPixelBar declaration

* fix error on plasmaeffect function

gustavoalara committed:
Added Counter Startup Message

gustavoalara committed:
Forgot to include common definitions

gustavoalara committed:
Open fire dev effect buttons (#4)

* Enables/Disables effects when press buttons

* Commandt to enable/disable effects on buttons

* Added button effect

* Added button effects

* Added button effects

* Added Effects to Buttons

* Added Effects to Buttons

* rolled back the previous change

* rolled back the previous change

* Added Effects to Buttons

* get current effect declaration

* getCurrentEffect function

* reworked buttons to effects functions

* fixed buttons names

* forgot to remove invalid function

* fucking bad memory

* added held variable

* rolled back (bad interpretation)

* fixed bad interpretations

* test data

* moron again

* Fixed Test data in the incorrect place

* moved checkAllButtonEffects to second core

* Improved leds redraw

* Improved leds redraw

* fix errors

* another little fix

* added effect origin detection

* added effect source detection

* added effect source detection

* added effect source detection

* fucking little finger

* Fixed  changes on variable names

* fixed error with activeButton condition

The condition was using the old varaiable name

* Added variables to remember the last serial effect

* Modified function to detect last serial effect

* Modified function to detect last serial effect

* Modified function to detect last serial effect

* creating public declarations and getters

* Modified function to detect last serial effect

* Modified function to detect last serial effect

* Modified function to detect last serial effect

* fixed duplication of variables

* removed test data

* Consolidated names

* little check when receiving new settings

* Added test data again for debug

* Consolidate button effects names due to overflow

* Consolidated button effect names due to overflow

gustavoalara committed:
Updated so that RGB color commands on the LED strip are applied only to the effects segment

gustavoalara committed:
Updated so that RGB color commands on the LED strip are applied only to the effects segment

gustavoalara committed:
Updated so that RGB color commands on the LED strip are applied only to the effects segment.

gustavoalara committed:
Removed previous change as not needed

gustavoalara committed:
Added new segment for status in the Neopixel strip

gustavoalara committed:
NeoPixel Strip Auto-configuration Logic if user only sets total LEDs

gustavoalara committed:
rebuild default neopixel segments initialization

gustavoalara committed:
Forgot #endif

SeongGino commented:
Apologies for the delay, but given the scale of this PR it'll take more time than usual for me to review. :heart:

gustavoalara committed:
Added Status segment to SetLedPackedColor function

gustavoalara commented:
Fantastic! I really appreciate that these changes are being accepted! And please, @SeongGino there’s no need to apologize for the delay! 

 I made the modifications you well suggested regarding the names for the definitions related to the 7-segment counter… but while making those changes, I also added a 5th segment for the NeoPixel LED strips. In this case, I added the possibility for the status signals (off-screen, docked mode, etc.) to also have their own segment.

gustavoalara committed:
Update .gitmodules

gustavoalara committed:
Update arduino-cli-builds.yml

gustavoalara committed:
Update arduino-cli-builds.yml

gustavoalara committed:
Update arduino-cli-builds.yml

gustavoalara committed:
Update submodule boards to latest commit

gustavoalara commented:
Hi @SeongGino, 

Any news about this pull request I made? Almost three months has passed and I didn't receive any question or request change about it. 

Thanks in advanced
